### PR TITLE
Feature/contrastive

### PR DESCRIPTION
**Pull request by benjijamorris**
## What does this PR do?
- Add VICReg Contrastive learning

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: configs/data/contrastive/vicreg.yaml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+      - _target_: monai.transforms.ToTensord
+        keys: image
+        track_meta: false
+      - _target_: monai.transforms.NormalizeIntensityd
+        keys: image
+
+      - _target_: monai.transforms.CopyItemsd
+        keys: image
+        times: 1
+        names: ["image_aug"]
```
Comment (ritvikvasan) [inline configs/data/contrastive/vicreg.yaml:25]: these names could also go in the aux section?

File: configs/data/contrastive/vicreg.yaml [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  batch_size: 128
+
+train_dataloaders:
+  _target_: monai.data.dataloader.DataLoader
+  batch_size: ${data._aux.batch_size}
+  num_workers: 8
+  shuffle: true
+  persistent_workers: false
+  dataset:
+    _target_: cyto_dl.datamodules.torchvision.MNIST
```
Comment (ritvikvasan) [inline configs/data/contrastive/vicreg.yaml:11]: maybe relevant for future - AICS "MNIST" - https://open.quiltdata.com/b/allencell/tree/aics/aics_mnist/

File: configs/data/contrastive/vicreg.yaml [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+      - _target_: monai.transforms.RandFlipd
+        keys: ["image_aug"]
+        prob: 0.5
+        spatial_axis: 1
+      - _target_: monai.transforms.RandShiftIntensityd
+        keys: ["image_aug"]
+        prob: 1.0
+        offsets: 0.5
+      - _target_: monai.transforms.RandGaussianSmoothd
+        keys: ["image_aug"]
```
Comment (ritvikvasan) [inline configs/data/contrastive/vicreg.yaml:66]: sucks that you have to repeat the block for image_aug, but I don't know a better way....

File: configs/data/contrastive/vicreg.yaml [hunk 4]
```diff
@@ -0,0 +1,136 @@
+_aux:
```
Comment (ritvikvasan) [inline configs/data/contrastive/vicreg.yaml:1]: should this data file be called mnist_vicreg.yaml?

File: configs/model/contrastive/vicreg.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
+_target_: cyto_dl.models.contrastive.Contrastive
+
+anchor_key: image
+positive_key: image_aug
+target_key: target
```
Comment (ritvikvasan) [inline configs/model/contrastive/vicreg.yaml:5]: what is the "target" key? Didn't see this in the data loader

File: cyto_dl/models/contrastive/contrastive.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        metrics = base_kwargs.pop("metrics", _DEFAULT_METRICS)
+        super().__init__(metrics=metrics, **base_kwargs)
+
+        self.backbone = backbone
+        self.task_head = task_head
+
+    def forward(self, x1, x2):
+        return self.backbone(x1), self.backbone(x2)
+
+    def plot_neighbors(self, embedding1, embedding2):
```
Comment (ritvikvasan) [inline cyto_dl/models/contrastive/contrastive.py:63]: these are callbacks?

File: cyto_dl/models/contrastive/contrastive.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        # plot positive embeddings in green
+        ax.scatter(embedding2[:, 0], embedding2[:, 1], c="green")
+
+        # draw lines between anchor and positive, anchor and negative
+        ax.plot([embedding1[:, 0], embedding2[:, 0]], [embedding1[:, 1], embedding2[:, 1]], "gray")
+
+        fig.savefig(Path(self.hparams.save_dir) / f"{self.current_epoch}_neighbors.png")
+        plt.close(fig)
+
+    def plot_classes(self, predictions, labels):
```
Comment (ritvikvasan) [inline cyto_dl/models/contrastive/contrastive.py:94]: callback?

File: cyto_dl/models/contrastive/contrastive.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+    def predict_step(self, batch, batch_idx):
+        x = batch[self.hparams.anchor_key].as_tensor()
+        embeddings = self.backbone(x)
+        preds = pd.DataFrame(
+            embeddings.detach().cpu().numpy(), columns=[str(i) for i in range(embeddings.shape[1])]
+        )
+        for key in self.hparams.meta_keys:
+            preds[key] = batch[key]
+        preds.to_csv(Path(self.hparams.save_dir) / f"{batch_idx}_predictions.csv")
```
Comment (ritvikvasan) [inline cyto_dl/models/contrastive/contrastive.py:140]: callback?

File: cyto_dl/nn/head/vic_reg.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        Parameters
+        ----------
+        dimensions
+            List of dimensions for projector network. Should start with output dimension of backone network. Subsequent dimensions should be ~4x larger following vicreg
+        """
+        super().__init__()
+        layers = []
+        for i in range(len(dimensions) - 1):
+            layers += [
+                nn.Linear(dimensions[i], dimensions[i + 1]),
```
Comment (ritvikvasan) [inline cyto_dl/nn/head/vic_reg.py:18]: use cyto_dl.nn.layers.MLP

File: cyto_dl/nn/losses/vic_reg.py [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+    def forward(self, x, y):
+        batch_size, num_features = x.shape
+
+        # view invarance loss
+        repr_loss = F.mse_loss(x, y)
+
+        x = x - x.mean(dim=0)
+        y = y - y.mean(dim=0)
+        # variance loss
+        std_x = torch.sqrt(x.var(dim=0) + 0.0001)
```
Comment (ritvikvasan) [inline cyto_dl/nn/losses/vic_reg.py:30]: why is this factor added here to the variance? to avoid sqrt of 0?

**Review by ritvikvasan — CHANGES_REQUESTED** [review]
nice addition!!

File: cyto_dl/nn/losses/vic_reg.py (hunk 10 above)
Comment (benjijamorris) [reply cyto_dl/nn/losses/vic_reg.py:30]: yeah exactly

File: configs/model/contrastive/vicreg.yaml (hunk 5 above)
Comment (benjijamorris) [reply configs/model/contrastive/vicreg.yaml:5]: the mnist dataloader pops out that the label as well

File: cyto_dl/models/contrastive/contrastive.py (hunk 6 above)
Comment (benjijamorris) [reply cyto_dl/models/contrastive/contrastive.py:63]: I'm kind of inclined to leave them as functions unless they seem generally useful

Comment (ritvikvasan) [reply cyto_dl/models/contrastive/contrastive.py:63]: cool

**Review by ritvikvasan — APPROVED** [review]
looks good!
